File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

'''
Algorithm: DFS
T: O(N)
S: O(N)
'''

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        depth = 0

        if not root:
            return depth
        
        q = [root]
        # dfs
        while q:
            num = len(q)
            
            while num > 0:
                node = q.pop(0)

                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
                num -= 1
            
            depth += 1
        
        
        return depth
            
        

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxDepth(TreeNode node, int depth){
        if(node == null)
            return depth - 1;
        
        return Math.max(maxDepth(node.left, depth + 1), maxDepth(node.right, depth + 1));
    }
    
    public int maxDepth(TreeNode root) {
        return maxDepth(root, 1);
    }
}